src/main/java/org/jenkinsci/test/acceptance/FallbackConfig.java

```diff
@@ -167,13 +180,63 @@ private WebDriver createWebDriver(TestName testName) throws IOException {
             }
             return new RemoteWebDriver(
                     new URL(u), //http://192.168.99.100:4444/wd/hub
-                    DesiredCapabilities.firefox());
-
+                    DesiredCapabilities.firefox()
+            );
         default:
             throw new Error("Unrecognized browser type: "+browser);
         }
     }
 
+    private WebDriver createContainerWebDriver(TestCleaner cleaner, String image, MutableCapabilities capabilities) throws IOException {
+        try {
+            Path log = Files.createTempFile("ath-docker-browser", "log");
+            LOGGER.info("Starting selenium container. Logs in " + log);
+
+            if (!IOUtil.isTcpPortFree(4444)) throw new IllegalStateException("Port 4444 is occupied");
+
+            Docker.cmd("pull", image).popen().verifyOrDieWith("Failed to pull image " + image);
+            // While this only needs to expose two ports (4444, 5900), it needs to be able to talk to Jenkins running
+            // out of container so using host networking is the most straightforward way to go.
+            String[] args = {"run", "-d", "--shm-size=2g", "--network=host", image};
+
+            ProcessInputStream popen = Docker.cmd(args).popen();
+            popen.waitFor();
+            String cid = popen.verifyOrDieWith("Failed to run selenium container").trim();
+
+            new ProcessBuilder(Docker.cmd("logs", "-f", cid).toCommandArray()).redirectErrorStream(true).redirectOutput(log.toFile()).start();
+
+            Closeable cleanContainer = new Closeable() {
+                @Override public void close() {
+                    try {
+                        Docker.cmd("kill", cid).popen().verifyOrDieWith("Failed to kill " + cid);
+                        Docker.cmd("rm", cid).popen().verifyOrDieWith("Failed to rm " + cid);
+                    } catch (IOException | InterruptedException e) {
+                        throw new Error("Failed removing container", e);
+                    }
+                }
+
+                @Override public String toString() {
+                    return "Kill and remove selenium container";
+                }
+            };
+            Thread.sleep(3000); // Give the container and selenium some time to spawn
```

Comment:
This seems a bit fragile. Isn't there a way to check if the container fully started? Maybe checking some logs entries?

Reply:
Another way is to poll for a 200 response for up to some time. That will save some time if it's faster than 3000